// eslint-plugin-nestjs/lib/parse-int-pipe/parse-int-pipe.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("../utils");
exports.message = {
    prefer: 'Prefer transform pipe `ParseIntPipe` in @Param decorator',
    transformed: 'Redundant coercing to number, parameter `{{name}}` must be a number',
};
exports.parseIntPipe = {
    create(context) {
        const parameters = [];
        const checkAndReport = (argument) => {
            const param = parameters.find(p => p.name === argument.name);
            if (!param) {
                return;
            }
            if (param.decorator.transformed) {
                context.report(argument, exports.message.transformed, { name: param.name });
            }
            else {
                context.report(argument, exports.message.prefer);
            }
        };
        return {
            CallExpression: (node) => {
                if (isCoercingToNumber(node)) {
                    const [argument] = node.arguments;
                    checkAndReport(argument);
                }
            },
            UnaryExpression: (node) => {
                if (isCoercingToNumber(node)) {
                    checkAndReport(node.argument);
                }
            },
            FunctionExpression: (node) => {
                if (!Array.isArray(node.params)) {
                    return;
                }
                node.params
                    .map(param => ({ param, decorator: getDecorator(param) }))
                    .filter(x => x.decorator)
                    .forEach(({ param, decorator }) => {
                    parameters.push({ name: param.name, decorator });
                });
            }
        };
    }
};
function getDecorator(node) {
    // const decorator = (node.decorators || []).find(d => d.expression && d.expression.callee && d.expression.callee.name === 'Param');
    const decorator = utils_1.getDecoratorByName(node, 'Param');
    if (!decorator) {
        return;
    }
    const [, expr] = decorator.expression.arguments;
    return {
        transformed: Boolean(expr && expr.type === 'NewExpression' && expr.callee && expr.callee.type === 'Identifier' && expr.callee.name === 'ParseIntPipe'),
    };
}
function isCoercingToNumber(node) {
    if (node.callee) {
        if (node.callee.name === 'parseInt') {
            // const [, radix] = node.arguments;
            // if (radix && radix.type === 'Literal' && radix.value !== 10) {
            //     return false;
            // }
            return true;
        }
        if (node.callee.name === 'Number') {
            return true;
        }
        if (node.callee.object && node.callee.object.name === 'Number' && node.callee.property.name === 'parseInt') {
            return true;
        }
    }
    if (node.operator === '+' && node.prefix === true) {
        return true;
    }
    return false;
}
