// eslint-plugin-nestjs/lib/use-validation-pipe/use-validation-pipe.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("../utils");
exports.message = 'Use pipe `ValidationPipe` for data validation';
exports.useValidationPipe = {
    create(context) {
        let validationPipeInClass = false;
        return {
            ClassDeclaration: (node) => {
                const classUsePipes = utils_1.getDecoratorByName(node, 'UsePipes');
                validationPipeInClass = hasExprValidationPipe(classUsePipes);
            },
            'ClassDeclaration:exit': () => {
                validationPipeInClass = false;
            },
            MethodDefinition: (node) => {
                if (validationPipeInClass) {
                    return;
                }
                const param = node.value.params.find(p => utils_1.getDecoratorByName(p, 'Body') !== undefined);
                if (!param) {
                    return;
                }
                const body = utils_1.getDecoratorByName(param, 'Body');
                if (hasExprValidationPipe(body)) {
                    return;
                }
                const usePipes = utils_1.getDecoratorByName(node, 'UsePipes');
                if (hasExprValidationPipe(usePipes)) {
                    return;
                }
                const { typeAnnotation } = param;
                if (typeAnnotation && typeAnnotation.typeAnnotation && ['GenericTypeAnnotation', 'TSTypeReference'].includes(typeAnnotation.typeAnnotation.type)) {
                    context.report({ node: param, message: exports.message });
                }
            },
        };
    }
};
function hasExprValidationPipe(node) {
    if (!node) {
        return false;
    }
    return node.expression.arguments
        .some(argument => {
        return (argument.type === 'NewExpression' && argument.callee.type === 'Identifier' && argument.callee.name === 'ValidationPipe')
            || (argument.type === 'Identifier' && argument.name === 'ValidationPipe');
    });
}
